### ISSUE TITLE
[3.3.5] Combat trigger

### ISSUE THREAD

**Issue by immvp** [TrinityCore/TrinityCore#28103]
### Description

Combat currently gets triggered the same way for Caster and Target. 
If Caster casts Fireball, Caster will be put in Combat instantly, and so will the **target**. 

### Expected behaviour

Caster should be put in combat when projectile is in the air. Proof: https://youtu.be/fLjTRifi6dE?t=206 (03:26)
Target should be put in combat when projectile hits. Proof: https://youtu.be/Jug7beeLQ1A?t=367 (06:07, the guy literally zooms in on the thrown, also this is at 19 twinks so warriors can't charge in combat)

### Steps to reproduce the problem

1. Have Player A cast a projectile type ability at Player B
2. Observe Combat on Both players. They'll get Combat simultaneously.

### Branch

3.3.5

### TC rev. hash/commit

 TrinityCore rev. 5759246d73c7 2022-07-09 18:15:48 +0200 (main branch) (Unix, Release, Static)

### Operating system

Ubuntu 20.04

### Custom changes

Transmog Patch

**Comment by adeutscher** [comment]
Notes getting my bearings: 

```
Unit.h:1144: Unit::SetInCombatWith is a wrapper around CombatManager::SetInCombatWith
Unit.cpp:8284 Unit::EngageWithTarget: If target cannot have threat, set in combat. If it can have threat, add threat (which seems to invoke SetInCombatWith within it)
Spell.cpp:2317: Spell::TargetInfo::PreprocessTarget(Spell* spell) has a very specific call to CombatManager::SetInCombatWith for immunities
Spell.cpp:7621: Spell::DoEffectOnLaunchTarget: If the hit result is not an evade, the target is not friendly, (not a positive spell or is a dispel), spellInfo->HasInitialAggro() or unit->IsEngaged())
	IsEngaged (virtual method, Unit implementation is alias for IsInCombat)
```

`Spell.cpp:7621` looks like this:
```
// This will only cause combat - the target will engage once the projectile hits (in DoAllEffectOnTarget)
if (m_originalCaster && targetInfo.MissCondition != SPELL_MISS_EVADE && !m_originalCaster->IsFriendlyTo(unit) && (!m_spellInfo->IsPositive() || m_spellInfo->HasEffect(SPELL_EFFECT_DISPEL)) && (m_spellInfo->HasInitialAggro() || unit->IsEngaged()))
    m_originalCaster->SetInCombatWith(unit);
```

Recent history of this check:

* The comment was last modified 2018-01-03 with 532ab1c7f86
* The check was last modified 2018-02-10 with 45c5e1b9d63

Did a mock-build that just flat-out doesn't have the Spell.cpp:7621 check/engage at all. This had the following effect: 

* Projectile (Fireball/Shadowbolt/Throw): No combat for caster/target on launch, combat on impact
* Non-projectile (Slow): Immediate combat on 'impact', didn't notice a change

This suggests that the check at Spell.cpp:7621 is working as intended and CombatManager::SetInCombatWith needs some sort of nuance for one-sided combat.

**Comment by offl** [comment]
There's probably at least one open issue about this, maybe it has something useful but may be hard to find

**Comment by adeutscher** [comment]
I've got at least progress to a fix going with f05e593e3b. I learned that a PvP combat instance has a 'Suppress' method.

The effect of this is that only the caster is placed in combat when a spell launches. Like with the mock-build from my previous comment, the target is not placed in combat until they are hit.

One new new wrinkle that this causes is that is the target is not placed into combat at all if the spell misses. Is this intended behaviour as well, or do we need to set things so that a miss results in the target entering combat?

![image](https://user-images.githubusercontent.com/2055500/178174866-a6c78cc1-c213-40f0-a1ce-08854cabb207.png)

**Comment by adeutscher** [comment]
Similar to the miss, a reflected spell doesn't enter the reflecter into combat until the spell has impacted the original caster

**Comment by immvp** [comment]
Mised spells puts you in combat on TBC, it's gonna be impossible to find an example of someone missing a projectile in an old WOLTK Video, not gonna lie. @adeutscher

**Comment by adeutscher** [comment]
Yarr, that's what I was afraid of. Getting whiffed by an ambush isn't exactly the hottest thing to put in one's videos.

If nobody chimes in with a compelling reason that getting missed by a spell shouldn't put you in combat by tomorrow afternoon/evening, then I'll start digging into handling a miss. Once I find the right spot, I'll probably kick off by copying the check+method at `Spell.cpp:7621` as it was when I started.

**Comment by adeutscher** [comment]
Did a bit of investigation this evening. My first attempt was in `Spell::TargetInfo::DoDamageAndTriggers`, near the bottom where PvP flag updates are done. My testing failed here, either because I'm flat-out working in the wrong spot or I made a mistake in copying from the `Spell::DoEffectOnLaunchTarget` version of the check. I'll continue to investigate this in my next TC session.

I did write up the test cases that I'll be putting in the PR.

Setup:

1. On one account, create a character with a projectile spell. I went with a warlock for Shadow Bolt, but using Fireball would have been better
2. On another account, create a warrior to be a target for projectiles. Specifically making a warrior to test Spell Reflection in a test case.
3. Log into the characters, and put them in a setup that allows them to attack each other
4. Set up your windows so that you can conveniently watch the combat state of the caster and the target at once. 
5. Place the characters at maximum range for your spell.
6. Each of these tests assumes that the two characters are not in combat at step 1.

Test Cases:

1. Confirm that projectiles don't put the target into combat at launch (Hit).
    1. Cast a projectile at the target.
    2. While the projectile is in the air, observe that the caster is immediately placed into combat.
    3. While the projectile is in the air, observe that the target is not placed into combat.
    4. Confirm that the projectile cast has hit. If it misses, then pivot to Test 2. Test 2 is this test, but for the more unlikely case of non-hits
    5. Observe that the target is now in combat.
2. Confirm that projectiles don't put the target into combat at launch (Hit).
    1. Cast a projectile at the target.
    2. While the projectile is in the air, observe that the caster is immediately placed into combat.
    3. While the projectile is in the air, observe that the target is not placed into combat.
    4. Confirm that the projectile cast has missed. If it hits, then repeat until it does not
        * If you don't want to keep rolling the dice on a miss, then you could create a second caster with a large level distance
    6. Observe that the target is now in combat.
3. Confirm that reflected projectiles put the target in combat at time of reflection. Expecting this to act just like Test 2, but adding it here to cover our bases.
	1. Set the warrior up with a shield and teach them spell reflection
        * `.additem 32255`
        * `.learn 23920`
        * `.level 70`
    2. Set up enough Rage on the target, and cast Spell Reflection
    3. While Spell Reflection is active, cast a projectile at the target.
    4. While the projectile is in the air, observe that the caster is immediately placed into combat.
    5. While the projectile is in the air, observe that the target is not placed into combat.
    7. Confirm that the projectile cast has is reflected. If it does not, then repeat until it does
    8. Observe that the target is in combat BEFORE the spell projectile hits the caster.
    9. Confirm that the target is in combat after the reflected projectile hits the caster.

Test Results:

1. Success
2. So far, failed. Neither the caster or the target are being placed into combat.
3. Not attempted. Expecting this to fail until test 2 passes.

**Comment by adeutscher** [comment]
Separate note: Misses are not even flagging the for PvP combat. I was uncertain about combat flagging, but I'm pretty sure that PvP flagging on misses is supposed to be a thing.

**Comment by Jildor** [comment]
I also have another related situation about this, when a hunter's trap is triggered, affected target should put in combat, but owner of trap (hunter) shouldn't, currently both (hunter and target) enter in combat.
This is because originalcaster is hunter, but anyway, hunter shouldn't enter in combat with trap activation.

https://www.youtube.com/watch?v=RtkraBQbpgI&t=569s <-- min 9:29

https://www.warcraftmovies.com/movieview.php?id=160839 <-- min 4:00 and 9:17 (see water icon)

https://www.warcraftmovies.com/movieview.php?id=155267 <-- min 13:55

**Comment by adeutscher** [comment]
@Jildor I'm concerned that this similar flavour of bug has a much different location.

When I did my mock build without any of my PR changes and a commented-out combat-on-launch, the caster and the target were put in combat when the projectile hit.

I'll make the mock-build again off of the 3.3.5 branch. If this trap behaviour still happens off of the mock-build, then I think that the fix should be a separate issue, branch, and PR.

**Comment by adeutscher** [comment]
Traps. Oh boy, this took longer than I expected.

@Jildor These are my reads on your video examples, please let me know if I misread:

* Video 1: Freezing Trap does not apply combat to the hunter
* Video 2, part 1: Freezing trap did not seem to trigger combat on the victim. Hard to read because the hunter opened fire a split second afterwards.
* Video 2, part 2: Freezing trap did not put the hunter into combat. Easier to read, better freeze to shot timing.
* Video 3: Hunter definitely not placed into combat.

To confirm, my mock-build is commenting out this in `Spell::DoEffectOnLaunchTarget()` (line 7620 in 3.3.5):

```
if (m_originalCaster && targetInfo.MissCondition != SPELL_MISS_EVADE && !m_originalCaster->IsFriendlyTo(unit) && (!m_spellInfo->IsPositive() || m_spellInfo->HasEffect(SPELL_EFFECT_DISPEL)) && (m_spellInfo->HasInitialAggro() || unit->IsEngaged()))
        m_originalCaster->SetInCombatWith(unit);
```

These lines are also now removed from my PR branch because they were causing on-launch behaviour to regress to 3.3.5.

My results were not what I expected.

Mock-build off of 3.3.5:

* Frost Trap: Does not place in combat. However, it does flag an unflagged trapper for PvP.
* Explosive Trap: Does not place in combat. Was not looking for PvP triggering in this build, but I think I did it once without triggering PvP
* Freezing Trap: Does not place in combat. 
* Snake Trap: Does not place in combat. The snakes don't attack until the trapper is placed in combat (they will continue to attack if the attacker leaves combat)

3.3.5:

* Frost Trap: Does not place in combat.
    * Didn't test PVP-flagging this time, but I didn't change anything on that front.
    * Places target in combat on application. To clear combat, the victim must leave the trap area and wait 4s.
* Explosive Trap: If PvP flags are proper, places both players in combat.
* Freezing Trap: If PvP flags are proper, places both players in combat. Does not flag the trapper for PvP.
* Snake Trap: Does not place both players in combat. Snake behavior is unchanged.

The PvP flagging weirdness threw me for a loop. It distracted me and threw off my testing plans in the mock-build. Redid the mock build and confirmed my results with both players definitely being PvP flagged off the bat.

It looks like my mock-build "fixed" the issue by accident. I ended up removing the `Spell::DoEffectOnLaunchTarget()` call of `SetInCombatWith` in my current PR for this issue, so I wanted to test the behavior. 

`bug/28103/noncombat-launch` (both players are PvP flagged from the start):

* Freezing Trap: The trapper is placed in combat, the victim is not.
* Frost Trap: Both trapper and victim are placed in combat.
* Explosive Trap: The trapper is placed in combat, the victim is not.
* Snake Trap: Same behaviour. No combat, and snakes don't attack unless attacked.

One way or the other, I'm pretty sure we have a bug with non-PvP players being able to set off traps against PvP players. Unless this is Blizzlike? I've never played a hunter, but I feel like I still would have noticed this if it were...

The combat issue was fixed with my mock build. If my current PR branch for this issue goes through, then the offending combat enter is moved from `Spell.cpp:7621` to `Spell.cpp:2152`. It's a bit changed in that the victim doesn't get put into combat, but the trapper still is.

The fix to the in-combat issue will probably be closely related to the areas that I've been adjusting in my PR, but I think this might still be out of scope for the issue and the PR's intent. I'd be more comfortable making another issue for this. Will confirm with the person that I've been talking with on that PR.

**Comment by Jildor** [comment]
No, the owner of the trap must not enter/refresh combat with the hit of the trap spell that hits the target, but the target that is hit by the trap spell must enter/refresh combat.
You can check it in the second video that I put as proof: https://www.warcraftmovies.com/movieview.php?id=160839 <-- min 4:00 and 9:17

if you see unitframes bar, where the name of the player and the player's target is, you can see that when they enter combat, just to the right of the name appears the small icon of the two swords that indicates that they are in combat
you can see how the hunter doesn't get in combat and target of trap yes.

**Comment by Jildor** [comment]
I also wonder if the combat should be left at the same time by both the target and the caster with spells with projectiles that hit after a period of time or should have a separate time for each one, because with the PR that you are trying to do it seems to me that will end combat at the same time even if the caster has entered combat before because the projectile has taken a while to hit the target.

**Comment by adeutscher** [comment]
@Jildor I found the 4:00 part of that video hard to parse because the hunter opened fire a split-second after the trap hit. The other ones were straightforward, though.

I'm unsure if the combat-end timing is not working as intended. My thinking is that the projectile launch will put the caster in combat, and then the projectile impact will refresh the combat status of the caster.

Though they're both combat-related, I feel like combat-end timing and hunter traps are separate enough from this issue's original "projectiles shouldn't immediately cause target combat" scope that they should be separate issues/discussions.

**Comment by Jildor** [comment]
As I indicated before, you can check the min 9:17, it should work in the way I indicate, there is no discussion on my part
Main issue starts by "Combat currently gets triggered the same way for Caster and Target." so can be related anyway.

What I intend to indicate is that we must first analyze how it should work and then find the best solution.
By example, sanctuary effects like vanish, shadowmeld or invisibility that causes caster instant leave combat, in pvp doesn't do mutualpvp leave combat (engaged players with that caster should keep combat), that´s working fine if I remember right and here we have a proof: https://www.youtube.com/watch?v=4BSw1UjrECU min 4:35

Probably the part of leaving combat in case of entering with projectiles is probably the same, it shouldn't be mutual leave combat, but I haven't had time to find any videos.

But that makes me think that maybe they shouldn't get into mutual combat when cast any spell, and then if is delayed spell, suppress combat for the target and after delayed ends, target set in combat.
Perhaps should engage the caster when casting the spell and when the spell hits the target engage the target with the caster of the spell, regardless of whether the spell is delayed or not.
and that each one has its timer to get out of combat separately unless it dies, in which case it should suppress combat with all engaged targets.

**Comment by ex1dreamer** [comment]
sometimes combat trigger do not reset at all.
